File: raspberry/ml_label_classify.py
"""
Map raw ML / YOLO class names to backend /api/hardware/sort waste_type values.
Keep aligned with frontend/src/utils/mlLabelClassify.js and backend/utils/mlLabelClassify.js
"""

VALID = ("Recycle", "Non-Bio", "Biodegradable", "Unsorted")


def classify_raw_label_to_waste_type(raw: str) -> str:
    s = (raw or "").strip()
    if not s:
        return "Unsorted"
    upper = s.upper()
    spaced = " ".join(s.lower().split())
    compact = "".join(c for c in s.lower() if c.isalnum())

    bio_phrases = (
        "FOOD WASTE", "ORGANIC", "COMPOST", "BANANA PEEL", "FRUIT PEEL", "VEGETABLE PEEL",
        "LEAVES", "GRASS", "EGG SHELL", "COFFEE GROUND", "TEA BAG", "RICE", "BREAD", "MEAT",
    )
    if any(p in upper for p in bio_phrases):
        return "Biodegradable"
    for w in ("food", "fruit", "vegetable", "compost", "organic", "peel", "leaf"):
        if w in spaced or w in compact:
            if "PLASTIC" not in upper and "METAL" not in upper and "BATTERY" not in upper:
                return "Biodegradable"

    recycle_phrases = (
        "PLASTIC BOTTLE", "WATER BOTTLE", "PET BOTTLE", "SODA CAN", "METAL CAN", "ALUMINUM CAN",
        "GLASS BOTTLE", "GLASS JAR", "NEWSPAPER", "CARDBOARD", "PAPER CUP", "TETRA PAK",
    )
    if any(p in upper for p in recycle_phrases):
        return "Recycle"
    for t in ("plasticbottle", "bottle", "glass", "metalcan", "aluminum", "paper", "cardboard", "recycl"):
        if t in compact:
            return "Recycle"

    non_bio = (
        "HAIR CLIP", "HAIRCLIP", "DIAPER", "STYROFOAM", "BATTERY", "TOOTHBRUSH", "CD", "DVD",
        "PLASTIC FORK", "PLASTIC SPOON", "FOAM CUP",
    )
    if any(p in upper for p in non_bio):
        return "Non-Bio"
    if "PLASTIC" in upper and "BOTTLE" not in upper and "CUP" not in upper and "CONTAINER" not in upper:
        return "Non-Bio"

    legacy = {
        "recycle": "Recycle", "recyclable": "Recycle", "non-bio": "Non-Bio", "nonbio": "Non-Bio",
        "biodegradable": "Biodegradable", "bio": "Biodegradable", "unsorted": "Unsorted",
    }
    if spaced in legacy:
        return legacy[spaced]
    if compact in legacy:
        return legacy[compact]

    return "Unsorted"


def normalize_waste_type(value: str) -> str:
    v = (value or "").strip()
    if v in VALID:
        return v
    return classify_raw_label_to_waste_type(v)

File: frontend/src/utils/mlLabelClassify.js
/**
 * Map raw ML / YOLO class names (e.g. "HAIR CLIP", "plastic_bottle") to backend
 * /api/hardware/sort values: Recycle | Non-Bio | Biodegradable | Unsorted
 *
 * Keep in sync with backend/utils/mlLabelClassify.js
 */

const VALID = ['Recycle', 'Non-Bio', 'Biodegradable', 'Unsorted'];

/** @param {string} raw */
export function classifyMlLabelToWasteType(raw) {
  const s = String(raw ?? '').trim();
  if (!s) return 'Unsorted';

  const upper = s.toUpperCase();
  const spaced = s.toLowerCase().replace(/\s+/g, ' ').trim();
  const compact = s.toLowerCase().replace(/[^a-z0-9]/g, '');

  // --- Biodegradable first (avoid "apple" device etc.; use food/organic cues) ---
  const bioPhrases = [
    'FOOD WASTE',
    'FOOD SCRAPS',
    'ORGANIC',
    'COMPOST',
    'BANANA PEEL',
    'FRUIT PEEL',
    'VEGETABLE PEEL',
    'LEAVES',
    'GRASS',
    'EGG SHELL',
    'COFFEE GROUND',
    'TEA BAG',
    'RICE',
    'BREAD',
    'MEAT',
    'FISH BONE',
    'CHICKEN BONE',
    'NAPKIN',
    'TISSUE PAPER',
    'PAPER TOWEL',
    'WOOD STICK',
    'TOOTH',
  ];
  const bioWords = [
    'food',
    'fruit',
    'vegetable',
    'banana',
    'apple',
    'orange',
    'mango',
    'leaf',
    'leaves',
    'grass',
    'compost',
    'organic',
    'eggshell',
    'eggshells',
    'bread',
    'rice',
    'meat',
    'bone',
    'peel',
    'stale',
    'rotten',
  ];
  if (bioPhrases.some((p) => upper.includes(p))) return 'Biodegradable';
  if (bioWords.some((w) => spaced.includes(w) || compact.includes(w.replace(/\s/g, '')))) {
    if (!upper.includes('PLASTIC') && !upper.includes('METAL') && !upper.includes('BATTERY')) {
      return 'Biodegradable';
    }
  }

  // --- Recyclables (bottles, metals, paper, clean cardboard) ---
  const recyclePhrases = [
    'PLASTIC BOTTLE',
    'WATER BOTTLE',
    'PET BOTTLE',
    'SODA CAN',
    'METAL CAN',
    'ALUMINUM CAN',
    'TIN CAN',
    'GLASS BOTTLE',
    'GLASS JAR',
    'BEER BOTTLE',
    'WINE BOTTLE',
    'PAPER CUP',
    'CARDBOARD BOX',
    'NEWSPAPER',
    'MAGAZINE',
    'OFFICE PAPER',
    'ALUMINUM FOIL',
    'STEEL',
    'TETRA PAK',
    'CARTON',
    'BOTTLE CAP',
    'CAN TAB',
  ];
  const recycleTokens = [
    'plasticbottle',
    'waterbottle',
    'bottle',
    'petbottle',
    'glass',
    'metalcan',
    'aluminum',
    'steelcan',
    'tin',
    'paper',
    'cardboard',
    'newspaper',
    'magazine',
    'carton',
    'tetra',
    'recycl',
    'can',
  ];
  if (recyclePhrases.some((p) => upper.includes(p))) return 'Recycle';
  if (recycleTokens.some((t) => compact.includes(t))) {
    if (compact.includes('plasticbag') || upper.includes('PLASTIC BAG')) return 'Recycle';
    if (compact.includes('plastic') && (compact.includes('bottle') || compact.includes('cup') || compact.includes('container'))) {
      return 'Recycle';
    }
    if (compact.includes('bottle') || compact.includes('glass') || compact.includes('metal') || compact.includes('paper')) {
      return 'Recycle';
    }
  }

  // --- Non-biodegradable / residual ---
  const nonBioPhrases = [
    'HAIR CLIP',
    'HAIRCLIP',
    'HAIR PIN',
    'BOBBY PIN',
    'DIAPER',
    'STYROFOAM',
    'POLYSTYRENE',
    'FOAM CUP',
    'RUBBER',
    'SILICONE',
    'LEATHER',
    'CERAMIC',
    'PORCELAIN',
    'CIGARETTE',
    'TOOTHBRUSH',
    'CD',
    'DVD',
    'BATTERY',
    'PLASTIC TOY',
    'PLASTIC FORK',
    'PLASTIC SPOON',
    'PLASTIC STRAW',
    'DISPOSABLE',
    'SYNTHETIC',
    'TAPE',
    'GLUE STICK',
    'FABRIC SCRAP',
    'NYLON',
    'PVC',
  ];
  const nonBioTokens = ['hairclip', 'diaper', 'styrofoam', 'straw', 'toothbrush', 'battery', 'cd', 'dvd', 'foam'];
  if (nonBioPhrases.some((p) => upper.includes(p))) return 'Non-Bio';
  if (nonBioTokens.some((t) => compact.includes(t))) return 'Non-Bio';
  if (upper.includes('PLASTIC') && !upper.includes('BOTTLE') && !upper.includes('CUP') && !upper.includes('CONTAINER')) {
    return 'Non-Bio';
  }

  // --- Legacy / generic labels from older maps ---
  const legacy = {
    recycle: 'Recycle',
    recyclable: 'Recycle',
    'recyclable waste': 'Recycle',
    recyclables: 'Recycle',
    'non-bio': 'Non-Bio',
    nonbio: 'Non-Bio',
    'non biodegradable': 'Non-Bio',
    'non-biodegradable': 'Non-Bio',
    'non biodegradable waste': 'Non-Bio',
    biodegradable: 'Biodegradable',
    'biodegradable waste': 'Biodegradable',
    bio: 'Biodegradable',
    unsorted: 'Unsorted',
    unknown: 'Unsorted',
    other: 'Unsorted',
  };
  const fromLegacy = legacy[spaced] || legacy[compact];
  if (fromLegacy && VALID.includes(fromLegacy)) return fromLegacy;

  return 'Unsorted';
}
